File: python/src/utils/text.py
import torch


def format_tensors_side_by_side(t1: torch.Tensor, t2: torch.Tensor) -> str:
	r1, c1 = t1.shape
	r2, c2 = t2.shape

	if r1 != r2:
		return "Tensors have different row counts and cannot be formatted side by side"

	result = ""

	# Format the tensors side by side
	for i in range(r1):
		for j in range(c1):
			result += f"   {format(float(t1[i][j]), '.6f')}   "
		result += "|"
		for j in range(c2):
			result += f"   {format(float(t2[i][j]), '.6f')}   "
		result += "\n"

	return result

File: Go/src/utils/text.go
package utils

import (
	"fmt"
	"strings"

	"gonum.org/v1/gonum/mat"
)

// Helper function to format matrices side by side
func FormatMatricesSideBySide(m1, m2 *mat.Dense) string {
	r1, c1 := m1.Dims()
	r2, c2 := m2.Dims()

	if r1 != r2 {
		return "Matrices have different row counts and cannot be formatted side by side"
	}

	var result strings.Builder

	// Format the matrices side by side
	for i := 0; i < r1; i++ {
		for j := 0; j < c1; j++ {
			result.WriteString(fmt.Sprintf("   %4f   ", m1.At(i, j)))
		}
		result.WriteString("|")
		for j := 0; j < c2; j++ {
			result.WriteString(fmt.Sprintf("   %4f   ", m2.At(i, j)))
		}
		result.WriteString("\n")
	}

	return result.String()
}
